Correctly handle bindings for the same key with and without --release

orestisfl opened the pull request:
Actually #2733's title is different from what I am fixing. I'll reread the thread tomorrow but `bindcode --release` works fine currently.

I am fixing the issue described in https://github.com/i3/i3/issues/2733#issuecomment-360805836.
I also wrote a small explanation in the second commit message.

- [x] I have
    ```
    mode "resize" {
    ...
        bindsym r --release exec --no-startup-id ~/bin/draw-resize.sh $(xdotool getactivewindow), mode "default"
    }
    bindsym $mod+r mode "resize"
    ```
    in my config. The first time I $mod+r it works as intended. The second time I $mod+r the r `--release` bindsym gets triggered immediately. This does not happen w/ `next`. Need to investigate.
- [ ] Tests from https://github.com/i3/i3/pull/3179#issuecomment-374261240

Fixes #2733

orestisfl committed:
Reduce repetition in get_binding()

orestisfl committed:
Correctly handle bindings for the same mod key with and without --release

Before this commit, get_binding() exited on the first match without
marking the rest --release bindings with B_UPON_KEYRELEASE_IGNORE_MODS.

Similarly, once it found a --release binding during a KeyPress event it
would stop searching for a matching key press binding.

Example config, placing the --release line first will trigger the second
problem:

# i3 config file (v4)
bindsym Super_L exec notify-send "press"
# or
# bindcode 133 exec notify-send "press"
bindsym --release Super_L exec notify-send "release"
# or
# bindcode --release 133 exec notify-send "release"

Fixes #2733

orestisf1993 commented on src/bindings.c:
```diff
@@ -270,23 +271,26 @@ static Binding *get_binding(i3_event_state_mask_t state_filtered, bool is_releas
+            return bind;
+        } else {
+            /* Continue looping to mark needed B_UPON_KEYRELEASE_IGNORE_MODS. */
+            result = bind;
```
This will change the current behaviour: if multiple bindings for the same key are defined (which is an error) the last of them will be executed. Will fix that.

Airblader commented:
So if it doesn't fix the issue mentioned in the commit message, we should change the commit message (to a "relates to")? Otherwise it'll auto-close the issue.

orestisfl committed:
Check for B_UPON_KEYRELEASE_IGNORE_MODS with bindsyms

From 548d74015c50d7fae14bfb8bb1989acde5fc22ae:
> 1. press $mod, press x, release x, release $mod
> 2. press $mod, press x, release $mod, release x

case (2.) didn't work, now it should be fixed.

orestisf1993 commented:
What I mean is that `bindcode --release` works fine in `next`.

I reread the whole thread since my understanding of the issue was not 100% correct.

### `bindsym` + `bindsym --release`
https://github.com/i3/i3/issues/2733#issuecomment-294005259
```
bindsym $mod+b exec notify-send "press"
bindsym --release $mod+b exec notify-send "release"
```
works in `next`.

### `bindcode` + `bindcode --release` with non-mod keys
https://github.com/i3/i3/issues/2733#issuecomment-294077895
```
bindcode 66 exec notify-send "press"
bindcode --release 66 exec notify-send "release"
```
works in `next`.


### `bindcode` + `bindcode --release` with mod keys
https://github.com/i3/i3/issues/2733#issuecomment-360445055
```
bindcode 133 exec notify-send "press"
bindcode --release 133 exec notify-send "release"
```
doesn't work in `next`, fixed in this PR.
Reason: once 133 (Super_L) is pressed the first `bind` with `->release == B_UPON_KEYPRESS` is matched and the loop exits in
https://github.com/i3/i3/blob/a9512c6345ae66c658e215396be1278754471671/src/bindings.c#L290
the other `bind` with `->release == B_UPON_KEYRELEASE` is never changed to `B_UPON_KEYRELEASE_IGNORE_MODS`.
The mods don't match in
https://github.com/i3/i3/blob/a9512c6345ae66c658e215396be1278754471671/src/bindings.c#L256-L257
since it's a mod key bindcode (I think, this is pretty confusing).


### `bindsym` + `bindsym --release` with mod keys
```
bindsym Super_L exec notify-send "press"
bindsym --release Super_L exec notify-send "release"
```
doesn't work in `next` but didn't work in the first version of this PR either, didn't notice.
Reason is that in this line
https://github.com/i3/i3/blob/a9512c6345ae66c658e215396be1278754471671/src/bindings.c#L239
`B_UPON_KEYRELEASE_IGNORE_MODS` is not checked.
This caused another bug. From 548d74015c50d7fae14bfb8bb1989acde5fc22ae:
> 1. press $mod, press x, release x, release $mod
> 2. press $mod, press x, release $mod, release x

case (2.) didn't work, now it should be fixed. Is there an issue for this?

I don't feel really confident about merging this since bindings are not tested very well.

Airblader commented:
> I don't feel really confident about merging this since bindings are not tested very well, let's give it some time.

Thanks, I appreciate that.

Maybe @stapelberg can chime in here, the bindings are more his area. :-)

orestisf1993 commented:
I could start adding some more binding tests. Not just for the changes, the whole thing.

Airblader commented:
That'd be wonderful. Perhaps we do it in a separate PR?

stapelberg commented:
More tests are always appreciated :). I haven’t looked at the code in question, but I’d be fine merging a bindcode-related change at the beginning of the development cycle (which is where we are right now). Most users use bindsym, so potential breakages are fairly limited.

orestisf1993 commented:
Actually, `258-keypress-release.t` and `290-keypress-numlock.t` test a bunch of cases after all, don't know if there are many stuff missing unrelated to this PR.

Maybe:
1. Test different order from 548d74015c50d7fae14bfb8bb1989acde5fc22ae
2. Test duplicate bindsym/bindcode (+permutations). In all cases the first binding should be called and an error should be produced.
3. With
    ```
    set $mod Mod4 # super_l
    bindsym Super_L nop 1
    bindsym --release nop 2
    bindsym $mod + a nop 3
    bindsym --release $mod + a nop 4
    ```
   - $mod press => nop 1
   - $mod release => nop 2
   - $mod press => nop 1
   - a press => nop 3
   - a release => nop 4
   - $mod release => ???

orestisf1993 commented:
> I have
>```
>mode "resize" {
>...
>    bindsym r --release exec --no-startup-id ~/bin/draw-resize.sh $(xdotool getactivewindow), mode "default"
>}
>bindsym $mod+r mode "resize"
>```
>in my config. The first time I $mod+r it works as intended. The second time I $mod+r the r `--release` bindsym gets triggered immediately. This does not happen w/ `next`. Need to investigate.

ha! This can happen in `next` too but with `bindcode --release`. Try:

```
# i3 config file (v4)
set $mod Mod4
mode "resize" {
    bindcode --release 27 mode "default"
}
bindsym $mod+r mode "resize"
```

orestisfl committed:
Reset B_UPON_KEYRELEASE_IGNORE_MODS bindings when switching modes

With example config:
    mode "a_mode" {
        bindcode 27 --release mode "default"
    }
    bindsym $mod+r mode "a_mode"

The first time $mod+r is pressed "a_mode" is activated like normal. When
r (bindcode 27) is pressed to exit the mode:
- On the KeyPress event the corresponding bind->release is correctly
marked as B_UPON_KEYRELEASE_IGNORE_MODS.
- On the KeyRelease event the command 'mode "default"' is executed but
bind->release is still B_UPON_KEYRELEASE_IGNORE_MODS since they are only
reset on KeyPress events.
The second time $mod+r is pressed and "a_mode" is activated and when the
r key is released the 'mode "default"' is executed even though the mods
are not matching since bind->release == B_UPON_KEYRELEASE_IGNORE_MODS.

This still doesn't catch 2 cases:
1. When the order is: press $mod -> press r -> release $mod -> release
    r. Since 'r' is released without any modifiers the binding matches.
2. With:
        mode "resize" {
            bindsym --release r mode "default"
        }
        bindsym r mode "resize"
    This is arguably correct: on the KeyPress event we switch to the mode and
    on the KeyRelease we switch back.

Airblader commented:
@orestisf1993 What's the status here?

orestisf1993 commented:
I've been running this patch for a while, seems fine. I think we can review this.
